### PR TITLE
fix[lang]: fix panic in call cycle detection

### PR DESCRIPTION
**Pull request by cyberthirst**
### What I did
- fix: https://github.com/vyperlang/vyper/issues/4194

### How I did it
- check if the func called is the func itself

### How to verify it
- added test

### Commit Message
```
fix cycle detection in `_compute_reachable_set()` by adding a check for
subcycles in the current call path

the function was detecting cycles by checking for a cyclic call to the
*root* of the call path

if the cycle was within the call path (excluding the root) it could
fall into infinite recursion
```

### CODE REVIEW

File: vyper/semantics/analysis/module.py [hunk 1]
```diff
@@ -161,6 +161,10 @@ def _compute_reachable_set(fn_t: ContractFunctionT, path: list[ContractFunctionT
             message = " -> ".join([f.name for f in path])
             raise CallViolation(f"Contract contains cyclic function call: {message}")
 
+        if g == fn_t:
```
Comment (charles-cooper) [inline vyper/semantics/analysis/module.py:164]: can we not merge this with the previous branch on line 160?

File: vyper/semantics/analysis/module.py [hunk 2]
```diff
@@ -150,15 +150,16 @@ def _compute_reachable_set(fn_t: ContractFunctionT, path: list[ContractFunctionT
...
 
     for g in fn_t.called_functions:
         if g in fn_t.reachable_internal_functions:
             # already seen
             continue
 
-        if g == root:
-            message = " -> ".join([f.name for f in path])
+        if g in path:
+            cycle_start = path.index(g)
```
Comment (charles-cooper) [inline vyper/semantics/analysis/module.py:160]: i think printing out the full call trace will be easier for the user to debug

File: vyper/semantics/analysis/module.py (hunk 1 above)
Comment (cyberthirst) [reply vyper/semantics/analysis/module.py]: opted for a different cycle detection approach: https://github.com/vyperlang/vyper/pull/4200/commits/5ac3ad037b4dcc023b2941609a0506b908b49b3b

File: vyper/semantics/analysis/module.py (hunk 2 above)
Comment (cyberthirst) [reply vyper/semantics/analysis/module.py]: addressed in: https://github.com/vyperlang/vyper/pull/4200/commits

File: vyper/semantics/analysis/module.py [hunk 3]
```diff
@@ -150,15 +150,15 @@ def _compute_reachable_set(fn_t: ContractFunctionT, path: list[ContractFunctionT
...
 
     for g in fn_t.called_functions:
         if g in fn_t.reachable_internal_functions:
             # already seen
             continue
 
-        if g == root:
-            message = " -> ".join([f.name for f in path])
+        if g in path:
+            extended_path = path + [g]
```
Comment (charles-cooper) [inline vyper/semantics/analysis/module.py:160]: nice, very clean!

**Review by charles-cooper — APPROVED** [review]

**Pull request merged**